File: packages/match-engine/src/services/windowManager.ts
import { Redis } from 'ioredis';
import { Pool } from 'pg';
import { REDIS_KEYS } from '@kinetik/shared';

export class WindowManager {
  private activeWindows: Map<string, { id: string; startedAt: number; expiresAt: number }> = new Map();

  constructor(
    private redis: Redis,
    private pool: Pool,
  ) {}

  /**
   * Activate a flash window - moves participants into the matching pool
   */
  async activateWindow(windowId: string): Promise<void> {
    // Update window status in DB
    await this.pool.query(
      `UPDATE flash_windows SET status = 'active' WHERE id = $1`,
      [windowId],
    );

    // Get all participants
    const participants = await this.pool.query(
      `SELECT user_id FROM window_participants WHERE window_id = $1 AND is_active = TRUE`,
      [windowId],
    );

    // Add participants to Redis match pool
    if (participants.rows.length > 0) {
      const userIds = participants.rows.map((r: any) => r.user_id);
      await this.redis.sadd(REDIS_KEYS.MATCH_POOL(windowId), ...userIds);
      await this.redis.expire(REDIS_KEYS.MATCH_POOL(windowId), 3600);

      // Also add to individual H3 queues
      for (const { user_id } of participants.rows) {
        const user = await this.pool.query(
          `SELECT h3_index FROM users WHERE id = $1`,
          [user_id],
        );
        if (user.rows[0]?.h3_index) {
          await this.redis.sadd(REDIS_KEYS.H3_QUEUE(user.rows[0].h3_index), user_id);
        }
      }
    }

    // Track active window
    const expiresAt = Date.now() + 30 * 60 * 1000; // 30 minutes
    this.activeWindows.set(windowId, {
      id: windowId,
      startedAt: Date.now(),
      expiresAt,
    });

    console.log(`[WindowManager] Activated window ${windowId} with ${participants.rows.length} participants`);
  }

  /**
   * Close a flash window - clean up participants and queues
   */
  async closeWindow(windowId: string): Promise<void> {
    // Update DB status
    await this.pool.query(
      `UPDATE flash_windows SET status = 'closed' WHERE id = $1`,
      [windowId],
    );

    // Clean up Redis queues
    await this.redis.del(REDIS_KEYS.MATCH_POOL(windowId));
    await this.redis.del(REDIS_KEYS.WINDOW_PARTICIPANTS(windowId));
    await this.redis.del(REDIS_KEYS.WINDOW_STATE(windowId));

    // Remove from active tracking
    this.activeWindows.delete(windowId);

    console.log(`[WindowManager] Closed window ${windowId}`);
  }

  /**
   * Get number of active windows
   */
  getActiveWindowCount(): number {
    return this.activeWindows.size;
  }
}

File: packages/match-engine-py/src/services/window_manager.py
"""
Window manager — handles flash window activation / deactivation and
participant queue management for the match engine.
"""

import logging
import time
from typing import Optional

import asyncpg

from ..config import settings
from .redis_client import RedisClient

logger = logging.getLogger(__name__)


class WindowManager:
    """Manages flash window lifecycle and participant H3 queues.

    Mirrors the TypeScript WindowManager but uses asyncpg for DB queries
    and the RedisClient for cache operations.
    """

    def __init__(self, redis: RedisClient, pg_pool: asyncpg.Pool):
        self._redis = redis
        self._pg = pg_pool
        self._active_windows: dict[str, dict] = {}

    # ── Activation ──────────────────────────────────────────────

    async def activate_window(self, window_id: str) -> None:
        """Activate a flash window — move participants into the Redis match pool."""
        # Update window status in DB
        async with self._pg.acquire() as conn:
            await conn.execute(
                "UPDATE flash_windows SET status = 'active' WHERE id = $1",
                window_id,
            )

            # Fetch all active participants
            rows = await conn.fetch(
                "SELECT user_id FROM window_participants WHERE window_id = $1 AND is_active = TRUE",
                window_id,
            )

        participant_ids = [row["user_id"] for row in rows]

        if participant_ids:
            # Add all participants to the Redis match pool
            pool_key = f"match:pool:{window_id}"
            await self._redis.sadd(pool_key, *participant_ids)
            await self._redis.expire(pool_key, 3600)

            # Add to individual H3 queues
            for uid in participant_ids:
                h3_row = await self._pg.fetchrow(
                    "SELECT h3_index FROM users WHERE id = $1",
                    uid,
                )
                if h3_row and h3_row["h3_index"]:
                    await self._redis.sadd(f"h3:queue:{h3_row['h3_index']}", uid)

        # Track active window
        now = time.time() * 1000
        self._active_windows[window_id] = {
            "id": window_id,
            "started_at": now,
            "expires_at": now + settings.window_duration_minutes * 60 * 1000,
        }

        logger.info(
            "Activated window %s with %d participants",
            window_id, len(participant_ids),
        )

    # ── Close ───────────────────────────────────────────────────

    async def close_window(self, window_id: str) -> None:
        """Close a flash window — clean up queues and DB status."""
        async with self._pg.acquire() as conn:
            await conn.execute(
                "UPDATE flash_windows SET status = 'closed' WHERE id = $1",
                window_id,
            )

        # Clean up Redis keys
        await self._redis.delete(
            f"match:pool:{window_id}",
            f"window:participants:{window_id}",
            f"window:state:{window_id}",
        )

        self._active_windows.pop(window_id, None)
        logger.info("Closed window %s", window_id)

    # ── Status ──────────────────────────────────────────────────

    def get_active_window_count(self) -> int:
        return len(self._active_windows)

    def is_window_active(self, window_id: str) -> bool:
        return window_id in self._active_windows
